### PR TITLE
feat(j-s): Multiple Indictment Sub Types

### PR DESCRIPTION
**Pull request by gudjong**
# Multiple Indictment Sub Types

[Fleiri en ein undirtegund brots - gagnamódel, migration og aðlögun á núverandi virkni](https://app.asana.com/0/1199153462262248/1203242082057175/f)

## What

- Adds backend support for multiple indictment sub types.
- Adjusts the frontend to use new data structure but assume a single sub type.

## Why

- Preparing for client support of multiple sub types.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: apps/judicial-system/api/src/app/modules/case/dto/createCase.input.ts [hunk 1]
```diff
@@ -15,8 +16,8 @@ export class CreateCaseInput implements CreateCase {
   readonly type!: CaseType
 
   @Allow()
-  @Field(() => String, { nullable: true })
-  readonly indictmentSubType?: IndictmentSubType
+  @Field(() => GraphQLJSONObject, { nullable: true })
+  readonly indictmentSubTypes?: IndictmentSubtypeMap
```
Comment (oddsson) [inline apps/judicial-system/api/src/app/modules/case/dto/createCase.input.ts:20]: Please rename this to indictmentSubtypes (not ...SubTypes)

File: apps/judicial-system/api/src/app/modules/case/dto/updateCase.input.ts [hunk 2]
```diff
@@ -27,8 +27,8 @@ export class UpdateCaseInput implements UpdateCase {
   readonly type?: CaseType
 
   @Allow()
-  @Field(() => String, { nullable: true })
-  readonly indictmentSubType?: IndictmentSubType
+  @Field(() => GraphQLJSONObject, { nullable: true })
+  readonly indictmentSubTypes?: IndictmentSubtypeMap
```
Comment (oddsson) [inline apps/judicial-system/api/src/app/modules/case/dto/updateCase.input.ts:31]: indictmentSubTypes --> indictmentSubtypes

File: apps/judicial-system/api/src/app/modules/case/models/case.model.ts [hunk 3]
```diff
@@ -40,8 +40,8 @@ export class Case implements TCase {
   @Field(() => String)
   readonly type!: CaseType
 
-  @Field(() => String, { nullable: true })
-  readonly indictmentSubType?: IndictmentSubType
+  @Field(() => GraphQLJSONObject, { nullable: true })
+  readonly indictmentSubTypes?: IndictmentSubtypeMap
```
Comment (oddsson) [inline apps/judicial-system/api/src/app/modules/case/models/case.model.ts:44]: indictmentSubTypes --> indictmentSubtypes

File: apps/judicial-system/backend/migrations/20221111133751-update-case.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
+'use strict'
+
+module.exports = {
+  up(queryInterface, Sequelize) {
+    return queryInterface.sequelize.transaction(async (transaction) =>
+      queryInterface
+        .addColumn(
+          'case',
+          'indictment_sub_types',
```
Comment (oddsson) [inline apps/judicial-system/backend/migrations/20221111133751-update-case.js:9]: Subtype is one word, so this should be `indictment_subtypes`

File: apps/judicial-system/backend/src/app/modules/case/dto/createCase.dto.ts [hunk 5]
```diff
@@ -18,9 +22,9 @@ export class CreateCaseDto {
   readonly type!: CaseType
 
   @IsOptional()
-  @IsString()
-  @ApiPropertyOptional({ enum: IndictmentSubType })
-  readonly indictmentSubType?: IndictmentSubType
+  @IsObject()
+  @ApiPropertyOptional()
+  readonly indictmentSubTypes?: IndictmentSubtypeMap
```
Comment (oddsson) [inline apps/judicial-system/backend/src/app/modules/case/dto/createCase.dto.ts:27]: indictmentSubTypes --> indictmentSubtypes

File: apps/judicial-system/backend/src/app/modules/court/court.service.ts [hunk 6]
```diff
@@ -177,17 +178,25 @@ export class CourtService {
...
+          !indictmentSubTypes ||
+          !indictmentSubTypes[policeCaseNumbers[0]] ||
+          indictmentSubTypes[policeCaseNumbers[0]].length !== 0)
+      ) {
         throw 'Sub type is required for indictments'
       }
 
       // At this point we know that we have a valid sub type
-      const subType = (isIndictment ? indictmentSubType : type) as SubType
+      const subType = (isIndictment && indictmentSubTypes
```
Comment (oddsson) [inline apps/judicial-system/backend/src/app/modules/court/court.service.ts:197]: Consider writing a helper function to get the case subtype.

File: apps/judicial-system/backend/src/app/modules/court/court.service.ts [hunk 7]
```diff
@@ -177,17 +178,25 @@ export class CourtService {
...
+        ? indictmentSubTypes[policeCaseNumbers[0]][0]
+        : type) as SubType
```
Comment (oddsson) [inline apps/judicial-system/backend/src/app/modules/court/court.service.ts:199]: Subtype

File: apps/judicial-system/web/src/components/InfoCard/InfoCardActiveIndictment.tsx [hunk 8]
```diff
@@ -41,8 +41,11 @@ const InfoCardActiveIndictment: React.FC = () => {
         {
           title: formatMessage(m.offence),
           value: capitalize(
-            workingCase.indictmentSubType
-              ? indictmentSubTypes[workingCase.indictmentSubType]
+            workingCase.indictmentSubTypes &&
+              Object.entries(workingCase.indictmentSubTypes).length > 0
+              ? indictmentSubTypes[
+                  Object.entries(workingCase.indictmentSubTypes)[0][1][0]
```
Comment (oddsson) [inline apps/judicial-system/web/src/components/InfoCard/InfoCardActiveIndictment.tsx:47]: This is unclear to me. Please consider writing a helper function for this and explain the `[0][1][0]` thing with a comment

File: apps/judicial-system/web/src/components/InfoCard/InfoCardClosedIndictment.tsx [hunk 9]
```diff
@@ -48,8 +48,11 @@ const InfoCardClosedIndictment: React.FC = () => {
         {
           title: formatMessage(m.offence),
           value: capitalize(
-            workingCase.indictmentSubType
-              ? indictmentSubTypes[workingCase.indictmentSubType]
+            workingCase.indictmentSubTypes &&
+              Object.entries(workingCase.indictmentSubTypes).length > 0
+              ? indictmentSubTypes[
+                  Object.entries(workingCase.indictmentSubTypes)[0][1][0]
```
Comment (oddsson) [inline apps/judicial-system/web/src/components/InfoCard/InfoCardClosedIndictment.tsx:54]: This is unclear to me. Please consider writing a helper function for this and explain the [0][1][0] thing with a comment

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/Defendant.tsx [hunk 10]
```diff
@@ -253,11 +255,17 @@ const Defendant: React.FC = () => {
...
             }}
             value={
-              workingCase.indictmentSubType
+              workingCase.indictmentSubTypes &&
+              Object.keys(workingCase.indictmentSubTypes).length > 0
                 ? {
-                    value: IndictmentSubType[workingCase.indictmentSubType],
+                    value:
+                      IndictmentSubType[
+                        Object.entries(workingCase.indictmentSubTypes)[0][1][0]
```
Comment (oddsson) [inline apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/Defendant.tsx:263]: This is unclear to me. Please consider writing a helper function for this and explain the [0][1][0] thing with a comment

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/Defendant.tsx [hunk 11]
```diff
@@ -253,11 +255,17 @@ const Defendant: React.FC = () => {
...
+                      ],
                     label: capitalize(
-                      indictmentSubTypes[workingCase.indictmentSubType],
+                      indictmentSubTypes[
+                        Object.entries(workingCase.indictmentSubTypes)[0][1][0]
```
Comment (oddsson) [inline apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/Defendant.tsx:267]: This is unclear to me. Please consider writing a helper function for this and explain the [0][1][0] thing with a comment

File: apps/judicial-system/web/src/utils/validate.ts [hunk 12]
```diff
@@ -204,10 +204,12 @@ export const isDefendantStepValidIndictments = (
 ) => {
   const result =
     policeCaseNumbers.length > 0 &&
+    workingCase.indictmentSubTypes &&
+    Object.entries(workingCase.indictmentSubTypes).length > 0 &&
     !someDefendantIsInvalid(workingCase) &&
     validate([
       [workingCase.type, ['empty']],
-      [workingCase.indictmentSubType, ['empty']],
+      [Object.entries(workingCase.indictmentSubTypes)[0][1][0], ['empty']],
```
Comment (oddsson) [inline apps/judicial-system/web/src/utils/validate.ts:212]: This is unclear to me. Please consider writing a helper function for this and explain the [0][1][0] thing with a comment

**Review by oddsson — CHANGES_REQUESTED** [review]
Some comments

File: apps/judicial-system/web/src/components/InfoCard/InfoCardActiveIndictment.tsx (hunk 8 above)
Comment (gudjong) [reply apps/judicial-system/web/src/components/InfoCard/InfoCardActiveIndictment.tsx:47]: This is throwaway code, so I would rather not do that. The whole logic will change when the client implements multiple indictment sub-types.

File: apps/judicial-system/web/src/components/InfoCard/InfoCardClosedIndictment.tsx (hunk 9 above)
Comment (gudjong) [reply apps/judicial-system/web/src/components/InfoCard/InfoCardClosedIndictment.tsx:54]: This is throwaway code, so I would rather not do that. The whole logic will change when the client implements multiple indictment sub-types.

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/Defendant.tsx (hunk 10 above)
Comment (gudjong) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/Defendant.tsx:263]: This is throwaway code, so I would rather not do that. The whole logic will change when the client implements multiple indictment sub-types.

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/Defendant.tsx (hunk 11 above)
Comment (gudjong) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/Defendant.tsx:267]: This is throwaway code, so I would rather not do that. The whole logic will change when the client implements multiple indictment sub-types.

File: apps/judicial-system/web/src/utils/validate.ts (hunk 12 above)
Comment (gudjong) [reply apps/judicial-system/web/src/utils/validate.ts:212]: This is throwaway code, so I would rather not do that. The whole logic will change when the client implements multiple indictment sub-types.

File: apps/judicial-system/api/src/app/modules/case/dto/createCase.input.ts (hunk 1 above)
Comment (gudjong) [reply apps/judicial-system/api/src/app/modules/case/dto/createCase.input.ts:20]: Actually subtype and sub-type are both valid, but subtype is more common. Will change.

**Review by oddsson — APPROVED** [review]

**Pull request merged**